### PR TITLE
[PURR][#2009]Include citation information of dataset in DOI metadata record

### PR DESCRIPTION
**Pull request by kuang5**
The changes include:
1. Include citation in the DOI xml when the DOI is created in the publication submission workflow.
2. Include citation for the DOIs that are set through publications on the citation admin interface when citation record is created.
3. Change the DOI registration http request method from "PUT" to "POST" according to DataCite's update on their API document.

The details can be found on PURR ticket #2009

### CODE REVIEW

File: core/components/com_citations/helpers/records.php [hunk 1]
```diff
@@ -0,0 +1,410 @@
...
+	 * Constructor
+	 *
+	 * @param      object $db     Database
+	 * @param      array  $config Optional configurations
+	 * @return     void
+	 */
+	public function __construct()
+	{
+		$this->_db  = \App::get('db');
+		$this->_publication_ver_tbl= '#__publication_versions';
```
Comment (zooley) [inline core/components/com_citations/helpers/records.php]: Citations should not have a direct dependency on Publications. Interactions between extensions should be handled with plugins (that respond, for example, to an onAfterSave event).

File: core/components/com_citations/helpers/records.php [hunk 2]
```diff
@@ -0,0 +1,410 @@
...
+	 */
+	public function getDoiList($pubIDArr, &$pubWithMultiVer)
+	{
+		$doiArr = [];
+
+		foreach ($pubIDArr as $pubID)
+		{
+			$sql = "SELECT doi FROM $this->_publication_ver_tbl WHERE publication_id = " . $pubID;
+			$this->_db->setQuery($sql);
+			$result = $this->_db->loadObjectList();
```
Comment (zooley) [inline core/components/com_citations/helpers/records.php]: Avoid raw SQL where possible and use models for information retrieval.

File: core/components/com_citations/helpers/records.php [hunk 3]
```diff
@@ -0,0 +1,410 @@
...
+		return $citationArr;
+	}
+
+	/**
+	 * Get DOI metadata XML from DataCite
+	 *
+	 * @param      $doi
+	 * @return     string  XML or false
+	 */
+	public function getDoiXML($doi)
```
Comment (zweidner) [inline core/components/com_citations/helpers/records.php]: This method might serve a more general purpose in Components\Publications\Models\Doi... or perhaps the same task could be accomplished using Components\Publications\Models\Doi::runCurl()? (Not entirely sure on that one...)

File: core/components/com_citations/helpers/records.php [hunk 4]
```diff
@@ -0,0 +1,410 @@
...
+
+	/**
+	 * Include citation and update DOI metadata record on DataCite
+	 *
+	 * @param      $xml
+	 * @param      $citationInfo
+	 *
+	 * @return     string  XML or false
+	 */
+	public function updateDoiXML($xml, $citationInfo)
```
Comment (zweidner) [inline core/components/com_citations/helpers/records.php]: This doesn't look vastly different from Components\Publications\Models\Doi::dataciteMetadataUpdate(), would it be better to update that?

File: core/components/com_citations/helpers/records.php (hunk 1 above)
Comment (kuang5) [reply core/components/com_citations/helpers/records.php]: The onAfterSave is never triggered. I have installed the plugin and the status is active. Also have run the muse migration command.

File: core/components/com_citations/helpers/records.php (hunk 3 above)
Comment (kuang5) [reply core/components/com_citations/helpers/records.php]: The two functions server different purposes. The function _getDoiXML() retrieving the DOI metadata from DataCite, however, runCurl is used to register the metadata of DOI on DataCite.

File: core/components/com_citations/helpers/records.php (hunk 4 above)
Comment (kuang5) [reply core/components/com_citations/helpers/records.php]: updateDoiXML only includes the citation information in DOI metadata, however, dataciteMetadataUpdate updates all DOI's metadata.

**Pull request closed without merging**